sql/src/main/java/io/crate/executor/transport/ChainableAction.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.executor.transport;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+class ChainableAction<ReturnType> {
+
+    private final Supplier<CompletableFuture<ReturnType>> doSupplier;
+    private final Supplier<CompletableFuture<ReturnType>> undoSupplier;
+
+    ChainableAction(Supplier<CompletableFuture<ReturnType>> doSupplier,
+                    Supplier<CompletableFuture<ReturnType>> undoSupplier) {
+        this.doSupplier = doSupplier;
+        this.undoSupplier = undoSupplier;
+    }
+
+    @VisibleForTesting
```

Comment:
I'd remove the annotation, as this is designed to be called from another component